Improve ABM time budget handling.

lhofhansl opened the pull request:
Handling long running ABMs has been awkward.
We used to run all ABM on all active blocks, and if a run took more than 200ms delay future runs based on how much longer than 200ms we took. That lead to very irregular ABM handling and still very long script runtimes because we do not check until after we ran all ABM on all blocks.

This PR turns this around:
1. First it shuffles the current active blocks. That is to ensure that no block receives preferred treatment.
2. Then we run ABMs against against blocks until we reached the time budget - still 200ms.
3. At that point we stop processing for this run. The next round with shuffle again and start again.

That way we can smoother behavior, and a better spread of the load across the active blocks.
On average the behavior is still that the ABMs are run less frequently on the active blocks.
The key point is that we stop ABM as soon as we reach the budget.

Please have a look. I tested this with some heavy mods (NSSM, etc) and the behavior is much more smooth. No more long gaps of no scripts.

SmallJoker commented on src/serverenvironment.cpp:
```diff
@@ -1340,11 +1340,6 @@ void ServerEnvironment::step(float dtime)
-			if (m_active_block_interval_overload_skip > 0) {
-				ScopeProfiler sp(g_profiler, "SEnv: ABM overload skips");
-				m_active_block_interval_overload_skip--;
-				break;
```
Now it no longer needs to be breakable at L1342

lhofhansl replied:
Good point!

SmallJoker commented on src/serverenvironment.cpp:
```diff
@@ -1354,30 +1349,41 @@ void ServerEnvironment::step(float dtime)
+				if (time_ms > max_time_ms) {
+					warningstream<<"active block modifiers took "
+								 <<time_ms<<"ms (processed " << i << " of "
+								 <<output.size()<<" active blocks)"<<std::endl;
```
Needs style fixes. Missing spaces and an indention level too much.

lhofhansl commented:
Remove the now unneeded do while(0) loop and fixed the white spacing.

SmallJoker commented:
Thanks, the code looks good now. Will test soon.

paramat commented:
Seems a good approach.

lhofhansl commented:
Hmm... Seems I messed up the PR when trying to rebase.

lhofhansl commented:
Fixed.

Also switched to the recommended way to shuffle (create a random device, use that generate a seed for a pseudo random number generator, and then use that generator with the shuffle function).
random_shuffle has been deprecated.

This also makes it more explicit that each time we call shuffle we get a different random order, which is essential to this!

lhofhansl commented:
Sorry... When I removed the do...while(0) "loop" I forgot to add braces for the wrapping if-statement.
Fixed and I tested this exact version again... Works fine as before.

thomasrudin commented:
Thats quite a good idea :smile:
What about globalsteps?
Those can get quite heavy too (they account for over 90% of the CPU-Time on my public server)

lhofhansl commented:
We could do a similar thing there. Not sure if we can suspend the execution of global step as easily, though (with no effect on gameplay).
I'll have a look, but let's that in another PR.

lhofhansl commented:
GlobalStep does not lend itself to do this as easily. We'd have to decide (1) what are useful units after which we could suspend execution for a bit (2) and make sure that everything still works as the module author intended.

thomasrudin commented:
I'll test that under load and will give feedback...

One thing that comes to mind is technic's ABM-trickery:
* The switching station is registered first (https://github.com/minetest-mods/technic/blob/master/technic/machines/switching_station.lua#L230) and sets a flag on every connected machine (abm: interval=1, chance=1)
* Then there is a timeout check (https://github.com/minetest-mods/technic/blob/master/technic/machines/switching_station.lua#L464) that relies on being called _after_ the switching station.. (abm: interval=1, chance=1)

I'm guessing this will be an issue under heavy load, but we will see :smile:

nerzhul commented on src/serverenvironment.cpp:
```diff
@@ -1338,47 +1339,55 @@ void ServerEnvironment::step(float dtime)
+		std::vector<v3s16> output(m_active_blocks.m_abm_list.size());
+		std::copy(m_active_blocks.m_abm_list.begin(), m_active_blocks.m_abm_list.end(), output.begin());
+
+		std::random_device rd;
```
don't instanciate mt19937 & random_device on each step. Make them owned by the environment

lhofhansl replied:
Good point. I made the random_device (non-member) static as there is no need to access it elsewhere. mt19937 is now a member. Have a look.

lhofhansl replied:
They could also both be members, non-static or not.

lhofhansl commented:
@thomasrudin only the active (map) blocks are shuffled. The ABMs are not. There are no dependencies between blocks. So everything should continue to work.

lhofhansl commented:
Any more comments here - especially on how to handle the the random device and pseudo random generator?

thomasrudin commented:
Just fyi: i pulled the commits a few days ago and it worked well
with around 15-25 players on pandorabox.
No issues so far...

lhofhansl commented:
@thomasrudin Nice!
Did you see any improvements in jerkiness (i.e. fewer times where scripts just - including global steps - appear to be stuck)?
Did you see any of the 200ms warning in the logs? (if not, you'll likely not see a difference)

--

BTW. I'm also thinking that instead of hardcoding the budget to 200ms, it might be better to make it a fraction of `abm_interval`. Right now `abm_interval` is 1s, so we could make 1/5 of that.
With that in mind, we could reduce `abm_interval` to less than 1s, together with this change it would contribute even more to more smooth behavior.

lhofhansl commented:
Just tried that. Set the timebudget to 1/5 of `abm_interval` and then set `abm_interval` to 0.1s. I.e. the timebudget each step is only 20ms, but we're running it each 100ms. Some of the mobs' (i.e. the active objects) jerkiness went away.

Not saying we should generally do that, there is some constant overhead in collecting the AMBs, shuffling the active blocks, etc. But since `abm_interval` *is* configurable the rest of the system should adapt.

lhofhansl commented:
Made the time budget adapt itself. I do not plan any further work on this, this should be good to merge.

lhofhansl commented:
0971a14a

thomasrudin commented:
Sorry for necroposting, but here is a composition-graph from my public server:
<img src="https://i.imgur.com/JLSt7iA.png">

The ABM-improvement is sadly not very noticeable :( because the most time
is consumed in globalsteps...

Maybe this will help you debug stuff like this in the future: https://monitoring.minetest.land/d/YUpouLmWk/overview?orgId=1&refresh=5s

/end shameless plug :smile:

lhofhansl commented:
Thanks @thomasrudin . In your case most time is spent on globalstep. I don't see a way to limit the runtime there.
I'd review the mods you have loaded (as if that isn't an obvious statement :) )

Ideally all Lua scripts should run asynchronously, but that'd require most of Minetests's internals to be thread safe.

lhofhansl commented:
400ms globalstep *is* pretty terrible. :(

lhofhansl commented:
How do you measure globalstep? The same loop also runs code for active objects (mobs) could limit those too, but it would lead to jerky movement.

thomasrudin commented:
## Lua side

The mod just wraps the existing globalsteps here:
https://github.com/thomasrudin-mt/monitoring/blob/master/builtin/globalstep.lua#L36

Pretty hacky but works very well :smile:

## C++ side

I do started to monitor the engine internals in my branch:
https://github.com/thomasrudin-mt/minetest/tree/prometheus

The results are here (pg/map save/load, async calls/time, sent blocks, etc):
https://pandorabox.io/grafana/d/ifuy9KSZz/minetest-engine?orgId=1&refresh=5s

I do see some potential in the map saving and block sending code (currently sync with the globalstep)
And i implemented a map save queue to offload the db-related parts:
https://github.com/thomasrudin-mt/minetest/tree/async-pg-mapsave

The code currently runs "productive" with a significant reduction of the lag-spikes...

lhofhansl commented:
Nice!! I might have time to dig some more.
